### ISSUE TITLE
float32 image data getting clobbered after changing ndisplay

### ISSUE THREAD

**Issue by tlambert03** [napari/napari#612]
## 🐛 Bug

While viewing 32 bit arrays, I saw some unexpected behavior, and I'm curious whether this is something you all have observed before or indeed a bug.  The gist is that when I add a 3D `float32` array to the viewer, then either press the "toggle displayed dims" button or directly change `viewer.dims.ndisplay`, the image appears to disappear, and it doesn't come back when going back to a 2D view.  Looking directly at the data in `viewer.layers[0].data` shows that it's all been permanently changed to extremely small numbers.  This is not the case for any other data type. 

It has something to do with vispy... and one way I could fix the problem is just making a copy of the `_data_view` in the `VispyImageLayer._on_data_change` method prior to calling `self.node.set_data` ... for instance by adding 
```python
 if dtype == 'float32':
    data = data.copy()
```
to the else statement in following section: 
https://github.com/napari/napari/blob/85e6bd46a2b1be230cceef429316e41405aa6356/napari/_vispy/vispy_image_layer.py#L70-L74

I feel like I'm spamming you recently with PRs, so here's a test that you can add to `test_qt_viewer.py` to try it out locally (for me, it only fails on the `float32` parameter). If you see the same behavior and you'd rather me submit a PR, let me know.  (and I'd also be curious whether you can think of a better fix than making a copy of the array) 

```python
import pytest

@pytest.mark.parametrize(
    "dtype", ['int8', 'uint8', 'int16', 'uint16', 'float32']
)
def test_qt_viewer_data_integrity(qtbot, dtype):
    """Test that the viewer doesn't change the underlying array."""

    image = np.random.rand(10, 32, 32)
    image *= 200 if dtype.endswith('8') else 2 ** 14
    image = image.astype(dtype)
    imean = image.mean()

    viewer = ViewerModel()
    view = QtViewer(viewer)
    qtbot.addWidget(view)

    viewer.add_image(image.copy())
    datamean = viewer.layers[0].data.mean()
    assert datamean == imean
    # toggle dimensions
    viewer.dims.ndisplay = 3
    datamean = viewer.layers[0].data.mean()
    assert datamean == imean
    # back to 2D
    viewer.dims.ndisplay = 2
    datamean = viewer.layers[0].data.mean()
    assert datamean == imean
```

**Comment by sofroniewn** [comment]
This may relate to #575 which also seems to be selectively affecting float32 and we regard as needing an urgent fix. We havn't dug in to figure out what's specifically going on with the float32 yet though

**Comment by tlambert03** [comment]
ah.  definitely seems like the same issue.  maybe the `data.copy()` strategy I mention above could be a short term fix (I guess it would waste memory?), while we dig into the vispy code to see what might be going on.

**Comment by sofroniewn** [comment]
This has been closed by #613 using the copy strategy.
